# Week 1 · Topic 4 — Input, Casting and f-strings

**Boxes: INPUT and OUTPUT, properly**

After this notebook you have everything you need for today's mini-project.

> **Already know this?** Skip to the *Topic 4 Challenge*.

---
## 1. Asking for input

In [4]:
record_id = input("Record ID : ")
print("You entered:", record_id)

You entered: 


---
## 2. The one rule that catches everyone

> **`input()` always gives you text. Always. Even when the person types a number.**

Run the next cell and type `23.7`.

In [7]:
value = input("Value : ")

print("You typed:", value)
print("Its type is:", type(value))

You typed: 23.7
Its type is: <class 'str'>


So this would fail with a `TypeError` — Python received the *characters* `2`, `3`, `.`, `7`, not a number.

Uncomment the line and see for yourself.

In [17]:
print(value + 1)

24.7


---
## 3. Casting — converting a type

| Function | Turns something into |
|---|---|
| `int()` | whole number |
| `float()` | decimal number |
| `str()` | text |

In [10]:
text = "20"
print(text, type(text))

as_number = int(text)
print(as_number, type(as_number))

20 <class 'str'>
20 <class 'int'>


Usually done in the same line as the input. Read it **inside out**: ask → get text → convert → store.

In [14]:
value = float(input("Value : "))

print(value + 1)
print(type(value))

24.7
<class 'float'>


> **Use `float()` unless you are certain the number is whole.** `int("23.7")` fails with a `ValueError`.

### Your turn
Ask for a number of GB used and a total GB, convert both, and print how many GB are free.

In [19]:
used = float(input("GB used: "))
total = float(input("GB total: "))
free = total - used
percent_free = (free / total) * 100
print(f"Free space: {free} GB ({percent_free:.2f}%)")


Free space: 10.0 GB (50.00%)


---
## 4. f-strings

Put `f` before the quotes, then any variable in `{ }`.

In [20]:
record_id = "R-004"
value = 23.7

print(f"Record {record_id} has value {value}")

Record R-004 has value 23.7


You can calculate inside the braces too.

In [21]:
limit = 20
print(f"Over the limit by {value - limit}")

Over the limit by 3.6999999999999993


Forgetting the `f` is the most common slip. This prints the braces themselves:

In [23]:
print(f"Record {record_id}")

Record R-004


### Your turn
Using an f-string, print `srv-01 is 87% full` from two variables.

In [26]:
serverid = "srv-01"
print(f"{serverid} is 87% full")


srv-01 is 87% full


---
## 5. Formatting — making it look like a report

| Written | Shows | Means |
|---|---|---|
| `{value}` | `23.7` | as is |
| `{value:.2f}` | `23.70` | 2 decimal places |
| `{value:>10}` | `      23.7` | right-aligned in 10 spaces |
| `{value:>10.2f}` | `     23.70` | both |
| `{diff:+.1f}` | `+3.7` | always show the sign |

In [27]:
percent = (value / limit) * 100

print(f"{percent}")
print(f"{percent:.2f}")
print(f"{percent:>12.2f}")

118.5
118.50
      118.50


Right-alignment is what makes a report line up instead of looking like a mess. Compare these two.

In [28]:
difference = value - limit

print("--- without alignment ---")
print(f"Value : {value}")
print(f"Limit : {limit}")
print(f"Diff  : {difference}")

print()
print("--- with alignment ---")
print(f"Value : {value:>10.2f}")
print(f"Limit : {limit:>10.2f}")
print(f"Diff  : {difference:>+10.2f}")

--- without alignment ---
Value : 23.7
Limit : 20
Diff  : 3.6999999999999993

--- with alignment ---
Value :      23.70
Limit :      20.00
Diff  :      +3.70


### Your turn
Print `7 / 3` to exactly **three** decimal places.

In [29]:
print(f"{7/3:.3f}")


2.333


---
## Topic 4 Challenge

Ask the user for a hostname, GB used and GB total. Then print exactly this shape, aligned:

```
==============================
  DISK REPORT  -  srv-01
==============================
  Used        :      87.00 GB
  Total       :     120.00 GB
  Free        :      33.00 GB
  Percent used:      72.5 %
==============================
```

This is a rehearsal for the mini-project. Get it working now and the lab is straightforward.

In [30]:
hostname = input("Enter hostname: ")
used = float(input("Enter GB used: "))
total = float(input("Enter GB total: "))
free = total - used
percent_free = (free / total) * 100

print("=" * 50)
print(f"  DISK REPORT  -  {hostname}")
print("=" * 50)
print(f"   used     : {used:>7.2f} GB")
print(f"   total    : {total:>7.2f} GB")
print(f"   free     : {free:>7.2f} GB")
print(f"   percent free : {percent_free:>5.1f}%")
print("=" * 50)

  DISK REPORT  -  srv-01
   used     :   90.00 GB
   total    :  126.00 GB
   free     :   36.00 GB
   percent free :  28.6%


---
## Same idea, three fields

Ask for something, convert it, report it back cleanly.

| Field | Looks like |
|---|---|
| **AI / Data Science** | `f"Mean: {mean:.2f} over {n} rows"` |
| **Cyber Security** | `f"[{level}] {failed} failed logins from {ip}"` |
| **IT** | `f"{hostname:<12} disk {used:>3} %"` |


---
## Common mistakes

| Mistake | What you see | Fix |
|---|---|---|
| `input("n: ") + 1` | `TypeError` | `float(input("n: ")) + 1` |
| `int(input())` when they type `23.7` | `ValueError` | use `float()` |
| `print("Value {v}")` | prints the braces | you forgot the `f` |
| `f"{value:.2}"` | odd result | you meant `.2f` |

## Go deeper

`examples/01_input_and_casting.py` — when to use `int()`, `float()`, or neither
`examples/02_fstrings.py` — decimal places, alignment, a finished report


---
## Recap

| You can now | Using |
|---|---|
| ask for a value | `input("...")` |
| convert it | `int()`, `float()`, `str()` |
| put values in text | `f"...{x}..."` |
| set decimal places | `{x:.2f}` |
| line a report up | `{x:>10.2f}` |

**You now have both boxes for Week 1.** Go and build your own in the lab.